### ISSUE TITLE
isOpen for TD/BU and possibly changing setLevel behavior

### ISSUE THREAD

**Issue by rvrolyk** [bujvary/hubitat#5]
Hey, sorry I missed something while looking at this earlier - `#isOpen` doesn't work when the blind is open because it's checking topPosition == 100 instead of 0.  For TD/BU fully open is bottom = 100 and top = 0.  The same problem actually exists in `#open` but the blind somehow just does the right thing there even though it's told to open the bottom all the way up and move the top all the way down.

The other thing I was wondering if you'd consider changing is #setLevel behavior.  HD doesn't expose individual blinds in Google Home so I'm using the Google Home Community app to do this which works great but it ends up having Open/Close mapped to #setLevel which works great in most cases except if you ask it to *open* or *close*.  In those cases it just moves the bottom rail even if the top was set to something unwanted.  I changed `#setLevel` to:

```
def setLevel(level, duration = null) {
    position = Math.min(Math.max(level.intValue(), 0), 100)
    def positions = []
    if (getShadeCapabilities() == 7 && device.currentValue('topPosition', true) != 0) {
        // The top rail is not closed but the bottom was told to move.  In nearly all
        // cases, this should mean the top closes.
        positions = [topPosition: 0, bottomPosition: level]
    } else {
        positions = [position: level]
    }
    parent.setPosition(device, positions)
}
```

and that seems to do the trick in that the top is always moved to the closed position which I think is generally what the user expects when they're moving the bottom (and it's how the remotes behave).  What are your thoughts?

**Comment by bujvary** [comment]
I've made the change to open() and isOpen() for TD/BU shades: bottom = 100 and top = 0

I also made the suggested change to setLevel().

Before I put out a new release I wanted to confirm that closed() and isClosed() are correct for TD/BU shades: bottom = 0 and top = 0

What's interesting is the latest version of the API document that I got from the developer at Hunter Douglas doesn't match what is actually happening with the TD/BU shade movement:

**ShadePosition**
Specifies the position of the shade. Top-down shades are in the same coordinate space as bottom-up
shades. Shade position values for top-down shades would be reversed for bottom-up shades. For
example, since 65535 is the open value for a bottom-up shade, it is the closed value for a top-down
shade. _The top-down/bottom-up shade is different in that instead of the top and bottom rail
operating in one coordinate space like the top-down and the bottom-up, it operates in two where
the top (middle) rail closed value is 0 and the bottom (primary) rail closed position is also 0 and
fully open for both is 65535._

**Comment by rvrolyk** [comment]
That is interesting.  Here's what I see --

Fully open as set from the app:
```
$ curl http://192.168.86.89/api/shades/44281
{"shade":{"id":44281,"type":8,"capabilities":7,"batteryKind":2,"smartPowerSupply":{"status":0,"id":0,"port":0},"batteryStatus":0,"batteryStrength":184,"roomId":31306,"firmware":{"revision":2,"subRevision":3,"build":3147,"index":3},"name":"QmVkcm9vbSAzIEJsaW5kIDE=","groupId":3225,"signalStrength":4,"positions":{"posKind1":1,"posKind2":2,"position1":65535,"position2":0}}}%
```

Fully closed:
```
$ curl http://192.168.86.89/api/shades/44281
{"shade":{"id":44281,"type":8,"capabilities":7,"batteryKind":2,"smartPowerSupply":{"status":0,"id":0,"port":0},"batteryStatus":0,"batteryStrength":184,"roomId":31306,"firmware":{"revision":2,"subRevision":3,"build":3147,"index":3},"name":"QmVkcm9vbSAzIEJsaW5kIDE=","groupId":3225,"signalStrength":4,"positions":{"posKind1":1,"posKind2":2,"position1":0,"position2":0}}}%
```

If I set that same blind to a top-down position and issue command to open it using the "fully open" value I see:
```
$ curl http://192.168.86.89/api/shades/44281
{"shade":{"id":44281,"type":8,"capabilities":7,"batteryKind":2,"smartPowerSupply":{"status":0,"id":0,"port":0},"batteryStatus":0,"batteryStrength":184,"roomId":31306,"firmware":{"revision":2,"subRevision":3,"build":3147,"index":3},"name":"QmVkcm9vbSAzIEJsaW5kIDE=","groupId":3225,"signalStrength":4,"positions":{"position2":14423,"position1":0,"posKind2":2,"posKind1":1}}}%

$ curl -vv -X PUT -H "Content-Type: application/json"  -d '{ "shade": { "positions": { "posKind1": 1, "position1": 65535, "posKind2": 2, "position2": 65535 } } }' http://192.168.86.89/api/shades/44281
*   Trying 192.168.86.89:80...
* Connected to 192.168.86.89 (192.168.86.89) port 80 (#0)
> PUT /api/shades/44281 HTTP/1.1
> Host: 192.168.86.89
> User-Agent: curl/7.77.0
> Accept: */*
> Content-Type: application/json
> Content-Length: 102
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< X-Powered-By: Express
< Content-Type: application/json; charset=utf-8
< Content-Length: 376
< Date: Thu, 20 Jan 2022 22:22:14 GMT
< Connection: keep-alive
<
* Connection #0 to host 192.168.86.89 left intact
{"shade":{"id":44281,"type":8,"capabilities":7,"batteryKind":2,"smartPowerSupply":{"status":0,"id":0,"port":0},"batteryStatus":0,"batteryStrength":184,"roomId":31306,"firmware":{"revision":2,"subRevision":3,"build":3147,"index":3},"name":"QmVkcm9vbSAzIEJsaW5kIDE=","groupId":3225,"signalStrength":4,"positions":{"posKind1":1,"position1":65535,"posKind2":2,"position2":65535}}}%
```

Visibly, this results in the top of the shade being fully open and the bottom being fully closed (eg the top rail is sitting on the bottom sill); I'm guessing in this case the top rail position "won" as the blind clearly looks as though the bottom didn't open.  If I interpret their doc a bit differently I think that makes sense - that is, they seem to be saying the coordinates for the top is  0 is closed and 65535 is open but for the top open is all the way down whereas the bottom is reversed even though the coordinates are the same.

In that context, a fully closed "window" would be top = 0 and bottom = 0.  A fully open "window" would be top = 0 (it needs to be up all the way) and bottom = 65535 *or* it could be top = 65535 and bottom = 0.  I suspect most people don't care about the second state (it's a bit of an odd one to be in).

**Comment by bujvary** [comment]
Thanks for the information. The new production version is available via HPM (v2.2.0) with the changes to open() and isOpen() as well as the suggested change to setLevel().

**Comment by rvrolyk** [comment]
awesome, thanks!
